### PR TITLE
Support secure and authenticated Milvus memory backends

### PR DESCRIPTION
**Pull request by chyezh**
### Background

I found that someone encounter some problem when using milvus as memory backend with zilliz cloud.
This PR will make easier to interact with zilliz cloud.

#1812 

### Changes

- add zilliz cloud to milvus memory backend implement.
- add document to readme.

### Documentation

- new feature and code is full documented.

### Test Plan

- no error occur when using the open-source milvus as memory backend.
- no error occur when using the zilliz cloud as memory backend.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: autogpt/memory/milvus.py
Comment (ngo275) [inline autogpt/memory/milvus.py]: https://github.com/Significant-Gravitas/Auto-GPT/issues/1812

I saw this post and was also working on the same fix on my local. I just needed to change `index_type` to `AUTOINDEX` and keep the rest the same as it used to be. 

```python
self.collection.create_index(
  "embeddings",
  {
    "metric_type": "IP",
    "index_type": "AUTOINDEX",
    "params": {"M": 8, "efConstruction": 64},
  },
  index_name="embeddings",
)
```

Comment (chyezh) [reply autogpt/memory/milvus.py]: metrics_type=IP is right.
params can be ignored by AUTOINDEX.
i have fixed it in new commit.

**Comment by chyezh** [comment]
I create a bad force-update after pull-request-merge from master.
Now, it's fixed and ready for review.

@ngo275 @richbeales

**Review by ngo275 — DISMISSED** [review]
I ran this version on my local to test Zilliz connectivity. It works well.

**Comment by Kenneth-zh** [comment]
I've configurate, but seemingly the code is running on default setting

**Comment by Kenneth-zh** [comment]
OK,it work perfectly

File: autogpt/memory/milvus.py
Comment (jaime0815) [inline autogpt/memory/milvus.py]: suggest using a common method to create the connection for milvus and zilliz, it is so wired to check whether has use_zilliz_cloud attribute within Milvus object.

Comment (chyezh) [reply autogpt/memory/milvus.py]: updated, general code in zilliz and milvus has been extract into init_collection function in new commit.

**Comment by chyezh** [comment]
fixed: black format check failed.

**Review by Kenneth-zh — APPROVED** [review]
It worked perfectly as I said

**Review by Kenneth-zh — APPROVED** [review]

**Review by Kenneth-zh — APPROVED** [review]

**Comment by xiaofan-luan** [comment]
Blocked by this issue, @Torantulino. Is there any way we can merge this as soon as possible? Thanks!

**Comment by chyezhz** [comment]
We found that many users use zilliz cloud to support Auto-GPT, but they have difficulties in using it. This PR can solve this problem. This is very important to us, hope this PR can be merged as soon as possible. 
@BillSchumacher @richbeales  thank you very much.

**Comment by chyezh** [comment]
it improves the use experience of Auto-GPT with zilliz cloud, 
can we merge this and close the related issue in shortly? Thanks.
@richbeales

**Comment by Torantulino** [comment]
We're taking a look at this, thanks!

File: autogpt/config/config.py
Comment (Pwuts) [inline autogpt/config/config.py]: `MEMORY_INDEX` already serves this purpose for multiple memory backends, and `MILVUS_COLLECTION` also already exists.

File: README.md
Comment (Pwuts) [inline README.md]: s/compatible/compatibility

File: .env.template
Comment (Pwuts) [inline .env.template]: Please remove (see main review comment https://github.com/Significant-Gravitas/Auto-GPT/pull/2127#pullrequestreview-1393644009)

File: .env.template
Comment (Pwuts) [inline .env.template]: Please move to below Milvus options

File: README.md
Comment (Pwuts) [inline README.md]: Please remove (see main review comment https://github.com/Significant-Gravitas/Auto-GPT/pull/2127#pullrequestreview-1393644009)

File: README.md
Comment (Pwuts) [inline README.md]: Please move to below Milvus

File: autogpt/memory/zilliz.py
Comment (Pwuts) [inline autogpt/memory/zilliz.py:1]: This looks like a generic `Milvus` with auth, is there anything in here specific to `Zilliz`?

**Review by Pwuts — CHANGES_REQUESTED** [review]
Thanks for submitting! It's nice to have options when it comes to memory backends (including SaaS-options), so in that sense it is a nice addition.

The implementation seems applicable to any authenticated Milvus instance: they use the same library, mostly the same implementation, except `ZillizMemory` uses authentication whereas `MilvusMemory` doesn't.

I suggest to remove the references to Zilliz in the implementation and merge the settings for Zilliz and Milvus as much as possible.
As far as I'm concerned, it can stay in the README, as a subheading or collapsible under **Memory Backend Setup > Milvus**.

Example:
<details>
<summary>Using a Zilliz cloud-based Milvus instance</summary>

* Docs about
* How to set up
* Zilliz with
* Auto-GPT
</details>

File: autogpt/config/config.py
Comment (chyezh) [reply autogpt/config/config.py]: Milvus's collection name do not accept character "-" of default `MEMORY_INDEX` set.
Should I modify the default MEMORY_INDEX setting to autogpt or keep the individual setting for milvus here.

File: autogpt/memory/zilliz.py
Comment (chyezh) [reply autogpt/memory/zilliz.py:1]: the index building parameters are also different.

**Comment by chyezh** [comment]
@Pwuts I have merged zilliz cloud support with milvus support as much as possible.
For showing correct memory backend type in command line, derived type ZillizMemory is still preserved.
image

File: autogpt/config/config.py
Comment (Pwuts) [inline autogpt/config/config.py]: * Preferably use empty defaults
* `ZILLIZ_URI` can be replaced by `MILVUS_URI`
```suggestion
        self.milvus_uri = os.getenv("MILVUS_URI")
        self.milvus_addr = os.getenv("MILVUS_ADDR", "localhost:19530")
        self.milvus_username = os.getenv("MILVUS_USERNAME")
        self.milvus_password = os.getenv("MILVUS_PASSWORD")
```

File: README.md
Comment (Pwuts) [inline README.md]:
```suggestion
#### Setup Client
```
same for the other subheadings

File: README.md
Comment (Pwuts) [inline README.md]: Thanks for adjusting, and sorry for requesting to change it again :')
```suggestion
  - [Milvus](https://milvus.io/) &ndash; [self-hosted](https://milvus.io/docs), or managed with [Zilliz Cloud](https://zilliz.com/)
```

File: .env.template
Comment (Pwuts) [inline .env.template]:
```suggestion
## milvus - Milvus (if configured - also works with Zilliz)
```

File: .env.template
Comment (Pwuts) [inline .env.template]: This is basically redundant. `Pymilvus` will determine the address as follows:
1. If `address` is set -> use that
2. If `address` is not set and `uri` is set, extract the `host:port` -> `address`
3. If `address` and `uri` are not set, use `DEFAULT_HOST:DEFAULT_PORT`

I'm fine with adding the URI config variable for convenience, and reversing the precedence of `address` and `uri`, but then name it `MILVUS_URI` and add a comment that `MILVUS_URI` takes precedence.
```suggestion
## MILVUS_ADDR - Milvus address (e.g. localhost:19530)
## MILVUS_URI - Milvus endpoint (e.g. https://xxx-xxxx.xxxx.xxxx.zillizcloud.com:443)
##   If set, MILVUS_URI overrides MILVUS_ADDR
```

File: README.md
Comment (Pwuts) [inline README.md]: s/fastly/quickly

s/or cloud service/or as a cloud service

File: .env.template
Comment (Pwuts) [inline .env.template]:
```suggestion
# MILVUS_ADDR=localhost:19530
# MILVUS_URI=
```

File: README.md
Comment (Pwuts) [inline README.md]: s/milvus/Milvus

File: README.md
Comment (Pwuts) [inline README.md]:
```suggestion
> ⚠️ If `MILVUS_URI` is set, it overrides `MILVUS_ADDR`
```

File: README.md
Comment (Pwuts) [inline README.md]:
```suggestion
- One of:
  - `MILVUS_ADDR=host:port` (for local instances)
  - `MILVUS_URI=https://xxx-xxxx.xxxx.xxxx.zillizcloud.com:19542`
```

File: README.md
Comment (Pwuts) [inline README.md]: These are optional, as they are not needed for local instances

**Review by Pwuts — CHANGES_REQUESTED** [review]
Thanks for revising! It looks better already, although there are a number of unresolved points. For one, I've dug a bit deeper into Pymilvus and Milvus, and it seems to me that there is no need for a specific implementation to use `Zilliz`.

* If Auto-GPT requires certain `index_parameters`, those should be the same regardless of where a Milvus instance is hosted, right?
* I agree that it's nice to be able to see what memory backend is being used, and whether it's a local or cloud-based instance if those options exist. It isn't part of *supporting* a memory backend though, so please create a separate PR to improve this bit of logging. E.g. give each backend a getter that returns relevant info about the type and the instance.
* The parameter `uri` of `connect()` is not specific to Zilliz, so the corresponding config variable should just be named `MILVUS_URI`.

Finally, the README could be cleaner. Maybe it helps to not consider Zilliz as the default use case.

**Comment by chyezh** [comment]
@Pwuts thank you for your patient review.
I agree with the most of the review suggestions and have implemented them in the new change. Some explanations about the review are as follows:

- About `index_parameters`:  because that Zilliz Cloud support `AutoIndex` to create index but `Milvus` not and Zilliz Cloud only support `AutoIndex`, the code handling `index_parameter` is still preserved. The unified behavior in `pymilvus` may be supported in the future, but now we need to add such a piece of code to keep compatibility for older version `pymilvus` in AutoGPT.
- Type `ZillizMemory` is also merged with `MivusMemory`, and the display of real memory backend support will be commit with separate PR in the future.

File: README.md
Comment (ashutoshpw) [inline README.md]:
```suggestion
- `MILVUS_USERNAME=your-username-of-milvus-or-cloud-service` # By default it is db_admin for Zilliz Cloud but can be different
```

File: autogpt/memory/milvus.py
Comment (ashutoshpw) [inline autogpt/memory/milvus.py]: @chyezh Will this work if there is no username and password? (In case of [Milvus Lite](https://milvus.io/docs/milvus_lite.md))

Comment (chyezh) [reply autogpt/memory/milvus.py]: It works well in milvus standalone, but test do not cover the milvus lite. username and password is empty string by default in connect function. it seems that set milvus_username milvus_password to empty string by default but not None is a better choice. I change it in new commit.

**Comment by chyezh** [comment]
@Pwuts @ashutoshpw 
- Rebase to latest master and pass the tests.
- `milvus_username` and `milvus_password` is set to empty string by default to keep consistent with `pymilvus`‘s default.

File: autogpt/memory/milvus.py
Comment (soothing-rain) [reply autogpt/memory/milvus.py]: > @chyezh Will this work if there is no username and password? (In case of [Milvus Lite](https://milvus.io/docs/milvus_lite.md))

It should work well with Milvus lite.

Milvus and Milvus Lite can share the same client code.

File: .env.template
Comment (jaime0815) [inline .env.template:102]: It seems that one configuration is enough for Milvus address/URI， and enables secure options if the URI contains https prefix before creating the connection.

Comment (chyezh) [reply .env.template:102]: I merge the URI and ADDR in new commit. it looks better now.

**Comment by chyezh** [comment]
@Pwuts All review has been reply.
Is there anything else need to be modify in these commit?
If not, can we merge the commit and close the related issue?
Looking forward to your reply.

**Comment by chyezh** [comment]
Resolve conflict, rebase to latest master, pass manually test.
But `codecov/patch` task not pass, how to fix it?

**Comment by ntindle** [comment]
You'll need to add some more tests that cover your changes. We have CodeCov configured not to let test coverage go down too far and your falls outside the threshold. It shouldn't be too bad just to cover the lines it says with tests if you click details on the failed run

File: autogpt/config/config.py
Comment (Pwuts) [inline autogpt/config/config.py]:
```suggestion
        self.milvus_username = os.getenv("MILVUS_USERNAME")
        self.milvus_password = os.getenv("MILVUS_PASSWORD")
```

File: autogpt/memory/milvus.py
Comment (Pwuts) [inline autogpt/memory/milvus.py]: `index_params` is determined based solely based on the url scheme here. If Zilliz requires specific `index_params`, they should be applied based on a hostname check.

File: autogpt/memory/milvus.py
Comment (Pwuts) [inline autogpt/memory/milvus.py]: After the check on line 18, this can still be a `http://` or `tcp://` URL, which is invalid for the `address` parameter

File: docs/configuration/memory.md
Comment (Pwuts) [inline docs/configuration/memory.md]:
```suggestion
#### In `.env` file
```

**Review by Pwuts — CHANGES_REQUESTED** [review]
This looks a lot better, thanks for applying the requested changes!

I have a few remaining requests, all minor things. :)

File: autogpt/config/config.py
Comment (chyezh) [reply autogpt/config/config.py]: The default `user` and `password` in pymilvus `connect` function is empty string.
image
I think that using empty string but not None as default value of `milvus_username` and `milvus_password` is reasonable.
Here's the previous suggestion of review.
https://github.com/Significant-Gravitas/Auto-GPT/pull/2127#discussion_r1173278808

**Comment by chyezh** [comment]
@Pwuts  Thank you for your feedback. Based on your review, I made the following fixes.
- Add `MILVUS_SECURE` for open-source milvus users to enable secure option.
- `connect` to `Milvus` is executed according to the following rules now.
    -  If your `MILVUS_ADDR` start with `https://` or `http://` or `tcp://`, it will use `uri` argument to connect, else use `address` argument.
    - If your `MILVUS_ADDR` start with `https`, it will enable `secure` option automatically.
    - If your `MILVUS_ADDR` meets the regex of zilliz cloud (provided by official), it will use AutoIndex automatically.

New commits covering more usage scenarios of milvus.

codecov/patch is not pass because of insufficient test coverage, could we merge the PR first and provide testing in future new PR. 
Many users use zilliz cloud to support Auto-GPT, but have trouble in running it. 
This is very important to us, hope this PR can be merged as soon as possible.

**Comment by ntindle** [comment]
I accept Codecov/patch failing for this PR. The team members who have been helping you are asleep but I’ve ping them to have a look.

File: autogpt/config/config.py
Comment (Pwuts) [reply autogpt/config/config.py]: You're right, passing `None` could break it.

**Comment by Pwuts** [comment]
Doing final review now

File: autogpt/config/config.py
Comment (Pwuts) [reply autogpt/config/config.py]: Fixed in 07e171c

**Comment by Pwuts** [comment]
PyMilvus indeed has empty strings as default `user`, `password` etc. However, this is of no concern to the user or main application config, so I amended `parse_configure` (now `configure`) to handle this (and to raise an error if one of them is set but not both). Also edited `__init__` to use one `connect()` statement as they were almost complete duplicates.

**Review by Pwuts — APPROVED** [review]

File: autogpt/memory/milvus.py
Comment (chyezh) [inline autogpt/memory/milvus.py:4]: `ModuleNotFoundError` is produced here in my python env.
`from autogpt.config import Config` is right.

Comment (Pwuts) [reply autogpt/memory/milvus.py:4]: And this is why we need unit tests :')

Can you submit a PR to fix? I'll merge it as soon as I see it.

Comment (chyezh) [reply autogpt/memory/milvus.py:4]: ok, I submit a new PR to fix it now.
It seems that `pymilvus` has been removed from requirement.
Unittest was skipped, I will submit a new PR to fix this problem in future.

Comment (chyezh) [reply autogpt/memory/milvus.py:4]: @Pwuts #3036 here :)

Comment (Pwuts) [reply autogpt/memory/milvus.py:4]: Please tag me when you submit a PR to fix the test :)

Comment (chyezh) [reply autogpt/memory/milvus.py:4]: OK

**Pull request merged**
